src/transforms.rs

```diff
@@ -0,0 +1,254 @@
+use jmespath::{
+    functions::{ArgumentType, CustomFunction, Signature},
+    Context, Expression, Rcvar, Runtime,
+};
+use rdkafka::Message;
+use serde_json::Value;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+#[derive(thiserror::Error, Debug)]
+pub enum TransformError {
+    #[error("Unable to mutate non-object value {value}")]
+    ValueNotAnObject { value: Value },
+
+    #[error("JmespathError: {source}")]
+    JmesPath {
+        #[from]
+        source: jmespath::JmespathError,
+    },
+
+    #[error("serde_json::Error: {source}")]
+    Json {
+        #[from]
+        source: serde_json::Error,
+    },
+}
+
+lazy_static! {
+    pub(crate) static ref TRANSFORM_RUNTIME: Runtime = {
+        let mut runtime = Runtime::new();
+        runtime.register_builtin_functions();
+        runtime.register_function("substr", Box::new(custom_substr()));
+        runtime
+    };
+}
+
+pub(crate) fn compile_transforms(
+    definitions: &HashMap<String, String>,
+) -> Result<HashMap<String, MessageTransform>, TransformError> {
+    let mut transforms = HashMap::new();
+
+    for (k, v) in definitions.iter() {
+        let t = match v.as_str() {
+            "kafka.partition" => MessageTransform::KafkaMetaTransform(KafkaMetaProperty::Partition),
+            "kafka.offset" => MessageTransform::KafkaMetaTransform(KafkaMetaProperty::Offset),
+            "kafka.topic" => MessageTransform::KafkaMetaTransform(KafkaMetaProperty::Topic),
+            "kafka.timestamp" => MessageTransform::KafkaMetaTransform(KafkaMetaProperty::Timestamp),
+            _ => {
+                let expression = TRANSFORM_RUNTIME.compile(v.as_str())?;
+                let expression = expression;
+
+                MessageTransform::ExpressionTransform(expression)
+            }
+        };
+
+        transforms.insert(k.to_owned(), t);
+    }
+
+    Ok(transforms)
+}
+
+fn custom_substr() -> CustomFunction {
```

Comment:
This function should be named to indicate what it's doing, since it's rather opaque it might be useful to throw a good number of comments around here :)

(I get that it's whatever this command line argument does)

I also think this function is worth some unit tests as well

Reply:
Renamed and commented. Will follow up with tests shortly.

Reply:
Added unit tests in https://github.com/delta-io/kafka-delta-ingest/pull/4/commits/fd7a7169c26fdbec90c7116e390fe88b9d3e7c71